### ISSUE TITLE
Shared Image reload() loses inital dimensions

### ISSUE THREAD

**Issue by wcout** [fltk/fltk#185]
I need to use Fl_Shared_Image::reload(), because the image file may have been replaced, or is even still the same image.
But after reload the data_w() and data_h() of the image do not have the dimensions from the file.

Have a look at this example (run from the test folder):

````c++
#include <FL/Fl.H>
#include <FL/Fl_Double_Window.H>
#include <FL/Fl_Button.H>
#include <FL/Fl_Box.H>
#include <FL/Fl_Shared_Image.H>

const char *name = "../documentation/src/Fl_File_Chooser.jpg";

Fl_Shared_Image *shi = 0;

static void cb_reload( Fl_Widget *wgt, void *d_ )
{
	if (shi && shi->original())
	{
		printf("before reload %s %dx%d (%dx%d)\n", shi->name(), shi->w(), shi->h(), shi->data_w(), shi->data_h());
		shi->reload();
		printf("after reload  %s %dx%d (%dx%d)\n", shi->name(), shi->w(), shi->h(), shi->data_w(), shi->data_h());
		Fl::first_window()->redraw();
	}
}

int main( int argc_, char *argv_[] )
{
	if ( argc_> 1 )
		name = argv_[1];
	fl_register_images();
	Fl_Double_Window win( 400, 300 );
	Fl_Box b( 110, 20, 170, 200 );
	b.color(FL_WHITE);
	b.box(FL_FLAT_BOX);
	Fl_Button but(10, 250, 100, 30, "reload");
	but.callback(cb_reload);
	shi = Fl_Shared_Image::get( name );
	if ( !shi )
		return 0;
	printf("image %s %dx%d (%dx%d)\n", shi->name(), shi->w(), shi->h(), shi->data_w(), shi->data_h());
	shi->scale( b.w(), b.h(), true, true);
	printf("image scaled %s %dx%d (%dx%d)\n", shi->name(), shi->w(), shi->h(), shi->data_w(), shi->data_h());
	b.image( shi );
	win.end();
	win.show();
	return Fl::run();
}
````

After reloading the initial dimensions are lost for ever.
I'm a little confused if this is to be expected or if it is an error.

**Comment by wcout** [comment]
I believe this change fixes the issue:
[shi_reload.diff.txt](https://github.com/fltk/fltk/files/5956165/shi_reload.diff.txt)

Resizable version of the exampe program:
```c++
#include <FL/Fl.H>
#include <FL/Fl_Double_Window.H>
#include <FL/Fl_Button.H>
#include <FL/Fl_Box.H>
#include <FL/Fl_Shared_Image.H>

const char *name = "../documentation/src/Fl_File_Chooser.jpg";

Fl_Shared_Image *shi = 0;

class ImageBox : public Fl_Box
{
public:
	ImageBox(int x_, int y_, int w_, int h_, const char *l_=0) : Fl_Box(x_,y_,w_,h_,l_) {}
	virtual void draw()
	{
		if ( image() )
			image()->scale(w(), h(), true, true);
		Fl_Box::draw();
	}
};

static void cb_reload( Fl_Widget *wgt, void *d_ )
{
	if (shi && shi->original())
	{
		printf("before reload %s %dx%d (%dx%d)\n", shi->name(), shi->w(), shi->h(), shi->data_w(), shi->data_h());
		shi->reload();
		printf("after reload  %s %dx%d (%dx%d)\n", shi->name(), shi->w(), shi->h(), shi->data_w(), shi->data_h());
		Fl::first_window()->redraw();
	}
}

int main( int argc_, char *argv_[] )
{
	if ( argc_> 1 )
		name = argv_[1];
	fl_register_images();
	Fl_Double_Window win( 400, 300 );
	ImageBox b( 110, 20, 170, 200 );
	b.color(FL_WHITE);
	b.box(FL_FLAT_BOX);
	Fl_Button but(10, 250, 100, 30, "reload");
	but.callback(cb_reload);
	shi = Fl_Shared_Image::get( name );
	if ( !shi )
		return 0;
	printf("image %s %dx%d (%dx%d)\n", shi->name(), shi->w(), shi->h(), shi->data_w(), shi->data_h());
	shi->scale( b.w(), b.h(), true, true);
	printf("image scaled %s %dx%d (%dx%d)\n", shi->name(), shi->w(), shi->h(), shi->data_w(), shi->data_h());
	b.image( shi );
	win.end();
 	win.resizable(b);
	win.show();
	return Fl::run();
}
````

**Comment by Albrecht-S** [comment]
Thanks for the report. I can't say much to the facts because the `Fl_Shared_Image::scale()` method and its interaction with stored (shared) images is still a mystery to me. Your patch looks interesting but I'd need to find the time to look into the details (which is currently not affordable).

I believe @ManoloFLTK could perhaps help with the scale() method and its internal data and maybe confirm the patch. (?)

**Comment by ManoloFLTK** [comment]
@wcout: please, report if the fix committed at ed5854d is OK for you.

**Comment by wcout** [comment]
@ManoloFLTK: Thanks for working on this. Unfortunately your change does not have the desired result: It keeps the `data_w() `and `data_h()`, but also changes the current `w()` and `h()`.

Output of test program with my patch:

image ../documentation/src/Fl_File_Chooser.jpg 498x408 (498x408)
image scaled ../documentation/src/Fl_File_Chooser.jpg 170x139 (498x408)
before reload ../documentation/src/Fl_File_Chooser.jpg 170x139 (498x408)
after reload  ../documentation/src/Fl_File_Chooser.jpg **170x139** (498x408)

Output ot the test program with your patch:

image ../documentation/src/Fl_File_Chooser.jpg 498x408 (498x408)
image scaled ../documentation/src/Fl_File_Chooser.jpg 170x139 (498x408)
before reload ../documentation/src/Fl_File_Chooser.jpg 170x139 (498x408)
after reload  ../documentation/src/Fl_File_Chooser.jpg **498x408** (498x408)

You can better see the implications if you remove the lines
```c++
		if ( image() )
			image()->scale(w(), h(), true, true);
````
in the `ImageBox::draw()` method of my test program. ==> the image jumps to a bigger size after reloading.

**Comment by Albrecht-S** [comment]
@wcout @ManoloFLTK As I see it, Manolo's commit fixes the "original size" issue as requested (data_w() and data_h() are preserved) but the missing piece seems to be the rescale to the previous w() and h() (i.e. the actual size before reload()). Is this true?

**Comment by wcout** [comment]
Yes. And the crux is, that the call to `update()` at the end will set w() and h() which are setter for **both** size and data size.

**Comment by wcout** [comment]
Also I am not sure if the comparison

`if ((img->w() != data_w() && data_w()) || (img->h() != data_h() && data_h())) {`

is really correct. I think the original comparison was ok because it compared "same things".

**Comment by ManoloFLTK** [comment]
Shouldn't update() let the image in the same state as that after get()?
I would tend to believe it should, but I'm ready to read other views.

With ed5854d  and the resizable version of your program, the relaoded image draws with its full resolution
on a retina display, as desirable.

The mentionned comparison is correct because, when it's performed, img is guaranteed to have w()==data_w()
and h()==data_h().

**Comment by Albrecht-S** [comment]
Okay, it's not as easy as it looks. I think we need to define first what `Fl_Shared_Image::reload()` is **expected** to do. Docs are really terse:
> Reloads the shared image from disk.

That's all. When I read this I'd expect the image to acquire its original size as it is stored on disk, but ...

The code comment says:
> // Make sure the reloaded image is the same size as the existing one.

The old code in 1.3 (which doesn't have `data_w()` and `data_h()` yet) uses
> Fl_Image *temp = img->copy(w(), h());

to restore the **current** image size.

So, basically, there's no "right or wrong" - we need to **define and document the correct behavior** and then implement it. We should also take backwards compatibility into account before we document the new (extended) behavior.

I could even imagine to extend the 1.3.6 (next release) documentation like:

"Reloads the shared image from disk _**and scales it to the current image size**_."

... because that's what it actually does. What would be the correct docs for 1.4 ???

**Comment by Albrecht-S** [comment]
> ... the crux is, that the call to update() at the end will set w() and h() which are setter for both size and data size.

@wcout I think these are only technical details and they must be called internally to set the correct values. They do not _change_ the image size.

So, technically spoken, we could still call `scale(old_w, old_h)` after calling `update()` if we saved the old w/h values before. Would that solve the issue?

**Comment by Albrecht-S** [comment]
> we could still call scale(old_w, old_h) after calling update() if we saved the old w/h values

Note: that would likely have to be `scale(old_w, old_h, 0, 1)` as proposed by @wcout in his patch.

**Comment by Albrecht-S** [comment]
One more thought: as I wrote above my expectation of the behavior of `reload()` would be different than what was implemented in 1.3.x. I can imagine keeping this behavior and extending it as appropriate as a backwards compatible default implementation and to add one or more optional parameters to let the user choose how it should behave, particularly arguments 3 and 4 used in above mentioned `scale(old_w, old_h, 0, 1)` call. One possible solution:

`Fl_Shared_Image::reload(int width, int height, int proportional=1, int can_expand=0);`

The parameters would be the same as scale() uses now.

**Comment by wcout** [comment]
> With ed5854d and the resizable version of your program, the reloaded image draws with its full resolution
on a retina display, as desirable.

Yes that's true, but please read my comment above on this commit regarding w() and and h(), and modify the test program to not scale the image in the draw() routine, to see what I mean.

> The mentionned comparison is correct because, when it's performed, img is guaranteed to have w()==data_w()
> and h()==data_h().

I don't think this is true, as can be seen with the test program:
`before reload ../documentation/src/Fl_File_Chooser.jpg 170x139 (498x408)`

**Comment by wcout** [comment]
@ManoloFLTK I'd be interested to know what are your objections to my patch. Do you think it is wrong?

**Comment by wcout** [comment]
>    ... the crux is, that the call to update() at the end will set w() and h() which are setter for both size and data size.

>> @wcout I think these are only technical details and they must be called internally to set the correct values. They do not change the image size.

But it will later on draw time scale the image to this w() and h().

**Comment by Albrecht-S** [comment]
@wcout I believe you misunderstood what I wanted to say.
`Fl_Shared_Image::update()` executes
>    w(image_->w());
>    h(image_->h());

which calls these methods of Fl_Image:
>  void w(int W) {w_ = W; data_w_ = W;}
>  void h(int H) {h_ = H; data_h_ = H;}

which are simple *assignments*, they do **not change** the image (size) which is at that point after Manolo's change the "requested" size, i.e. the current size of the image after the copy() operation. That size results from data_w() and data_h() of the current shared image which was loaded earlier.

**Comment by Albrecht-S** [comment]
BUT: I think I begin to understand the real underlying problem. In the old 1.3 code the image resizing could only be done by calling copy(data_w(), data_h()) which is what we still have. I suggested to call scale() internally after the copy() to resize (scale) the image, but now I think this is not correct.

INSTEAD: We should reload (i.e. "read the image from disk") and let its data size alone (_don't copy() it_) to keep the original sizes of the image as it is **now** (!) on disk. The only operation we should do is indeed a scale() operation which is ... exactly what @wcout suggested in his patch.

Example case: let's assume the original image was 100x100 scaled to 50x50, i.e. the original data size was 100x100. Let's further assume the image was **updated on disk** to a better resolution 200x200. As I understand the current code, `reload()` would do a `copy(100, 100)` which would discard the better resolution (data size 200x200) and create a new image with data size 100x100 and thus still "lose initial dimensions" of the **updated disk image**.

@ManoloFLTK Am I missing something?

**Comment by ManoloFLTK** [comment]
> > With [ed5854d](https://github.com/fltk/fltk/commit/ed5854d65a0b1da14eecf382e46512fbb9b94f02) and the resizable version of your program, the reloaded image draws with its full resolution
> > on a retina display, as desirable.
> 
> Yes that's true, but please read my comment above on this commit regarding w() and and h(), and modify the test program to not scale the image in the draw() routine, to see what I mean.
> 
> > The mentionned comparison is correct because, when it's performed, img is guaranteed to have w()==data_w()
> > and h()==data_h().
> 
> I don't think this is true, as can be seen with the test program:
> `before reload ../documentation/src/Fl_File_Chooser.jpg 170x139 (498x408)`

The image called img above refers to what is called img inside the code of Fl_Shared_Image::reload().
It's different from the image the test program shows the dimensions.

**Comment by Albrecht-S** [comment]
My test results:
```
// NEW CODE (Chris' patch)

image ../documentation/src/Fl_File_Chooser.jpg 498x408 (498x408)
image scaled ../documentation/src/Fl_File_Chooser.jpg 400x328 (498x408)
before reload ../documentation/src/Fl_File_Chooser.jpg 400x328 (data 498x408)
after reload  ../documentation/src/Fl_File_Chooser.jpg 400x328 (data 498x408) [1]
before reload ../documentation/src/Fl_File_Chooser.jpg 400x328 (data 498x408)
after reload  ../documentation/src/Fl_File_Chooser.jpg 400x328 (data 215x200) [2]

// Current code (Git ff0543524)

image ../documentation/src/Fl_File_Chooser.jpg 498x408 (498x408)
image scaled ../documentation/src/Fl_File_Chooser.jpg 400x328 (498x408)
before reload ../documentation/src/Fl_File_Chooser.jpg 400x328 (data 498x408)
after reload  ../documentation/src/Fl_File_Chooser.jpg 498x408 (data 498x408) [1]
before reload ../documentation/src/Fl_File_Chooser.jpg 498x408 (data 498x408)
after reload  ../documentation/src/Fl_File_Chooser.jpg 498x408 (data 498x408) [2]
```
Tests were done with the second demo program but with scaling inside draw() commented out as suggested by Chris (@wcout).

I consider the first part (Chris' patch) correct and the sencond part (Git current) faulty.

Test notes: The first four lines of each test were run as suggested with image Fl_File_Chooser.jpg.

IMHO the intended behavior at marker [1] is that the display size (w, h) keeps the same when the reload button is clicked. This is correct in the first part and wrong in the second part (current code).

After pressing the reload button for the first time I copied fl_color_chooser.jpg (size 215x200) over the original image and clicked the reload button again. The result is shown at marker [2]. Again, Chris' patch looks fine (data_size is 215x200) whereas the current Git version has 498x408 which is a copied and scaled version of the original image.

So far I'm coming to the conclusion that Chris' patch is superior and mostly correct for FLTK 1.4. The only point that needs to be discussed is how the scale() operation should be done. See also my previous comments.

**Comment by Albrecht-S** [comment]
I should add that I modified the window and image box sizes for my test: 800x600 and 400x400, resp.

**Comment by Albrecht-S** [comment]
Another **thought**: the issue here is partly caused by the old (FLTK 1.3 style) code of Fl_Shared_Image and new (FLTK 1.4 style) expectations. I looked closer at the Fl_Shared_Image code and I noticed that this is all based on the fact that images could have only one size and had to be copied to (re)scale them in FLTK 1.3.

Under these constraints Fl_Shared_Image::get(name, w, h) would try to find the image and *copy* it if not found directly to another size to scale it to (w, h).

In FLTK 1.4 though we have the scale() method which does no longer need a copy of the image to rescale it for drawing. I'm not sure about performance issues [1], but in comparison I'd say that Fl_Shared_Image could be rewritten (updated) to use scale() rather than copy() and this would, OTOH, make it almost obsolete because users can easily do this themselves in their program code. We could still keep its API but the changes under the hood would likely simplify the code pretty much (e.g. Fl_Shared_Image::get() and find()).

[1] ISTR and I assume that images are cached once they are drawn for the first time in both models which would save lots of memory for the no longer needed image copies of Fl_Shared_Images.

**Comment by ManoloFLTK** [comment]
I agree with the suggestion to keep the data size of the reloaded image and instead to change its drawing size.
This is now committed in the git repo.

About image caches:
In 1.4, as in 1.3, images are cached the 1st time they are drawn and the cache is re-used at each later redraws.
For Fl_RGB_Images, the cache has the size of the image data when the platform can simultaneously draw and scale :
- macOS can draw and scale,
- X11 + Render extension can draw and scale,
- Windows + AlphaBlend() from msimg32.dll can draw and scale to the display and always to printers.
But if the platform cannot draw-and-scale-an-image, the cache takes the drawing size of the image. That's
what is done under X11 in absence of the Render extension. When the screen is re-scaled, the effective drawing
size (img->w() * scale x img->h() * scale) changes, so all images are recached the first time they are redrawn.

The benefit of this is that images with large enough data size are always drawn to the full resolution of HighDPI
displays.

@Albrecht: is there any performance issue?

**Comment by Albrecht-S** [comment]
> For Fl_RGB_Images, the cache has the size of the image data when the platform can simultaneously draw and scale
> @albrecht: is there any performance issue?

@ManoloFLTK Thanks for you comment. After reading your explanation I don't think there's a performance issue if the caching is done effectively at the first draw operation. My concern **was** whether there's a difference if we scale the Fl_Shared_Image once by doing the `copy(w,h)` (your first commit, which would change the data size) vs. scaling it at draw time by setting only the display size with `scale(w,h)` (latest commit). The separate scaling to the display zoom factor you are referring to (if I understand correctly) is just another step that's common to both commits. Doing the copy() operation would also lose image quality, particularly if the current display size was lower than the original image resolution.

That said, the only performance issue I can see now is that on platforms that can "draw and simultaneously scale a cached image" there could be a performance penalty because the scaling to the screen factor is done every time an image is drawn vs. only once when the image is cached in the "zoomed" size. However, caching the image in its original resolution saves memory (for screen scaling factors > 1)and the system/platform drawing code should do its job well enough, so ... I think it's fine as it is.

**Comment by Albrecht-S** [comment]
@wcout Can you confirm the fix and can we (or you) close the issue now?

**Comment by wcout** [comment]
Glad you analysed and accepted my fix. Guess everybody learned something along the way (me: it's 'to lose' not 'to loose' for example ;))
Please feel free to close.

**Comment by Albrecht-S** [comment]
Thanks for confirmation.
